dask_match/core.py

```diff
@@ -616,7 +625,245 @@ def optimize_expr(expr):
+
+        # Traverse each "root" until we find a fusable sub-group.
+        # Here we use root to refer to a Fusable Expr node that
+        # has no Fusable dependents
+        roots = [
+            k
+            for k, v in dependents.items()
+            if v == set() or all(not isinstance(_expr, Fusable) for _expr in v)
+        ]
+        while roots:
+            root = roots.pop()
+            seen = set()
+            stack = [root]
+            group = []
+            while stack:
+                next = stack.pop()
+
+                if next._name in seen:
+                    continue
+                seen.add(next._name)
+
+                group.append(next)
+                for dep in dependencies[next]:
+                    if not (dependents[dep] - set(stack) - set(group)):
+                        # All of deps dependents are contained
+                        # in the local group (or the local stack
+                        # of expr nodes that we know we will be
+                        # adding to the local group)
+                        stack.append(dep)
+                    elif dep not in roots and dependencies[dep]:
+                        # Couldn't fuse dep, but we may be able to
+                        # use it as a new root on the next pass
+                        roots.append(dep)
+
+            # Replace fusable sub-group
+            if len(group) > 1:
+                group_deps = []
+                local_names = [_expr._name for _expr in group]
+                for _expr in group:
+                    group_deps += [
+                        operand
+                        for operand in _expr._subgraph_dependencies()
+                        if operand._name not in local_names
+                    ]
+                to_replace = {group[0]: FusedExpr(group, *group_deps)}
+                return replace_nodes(expr, to_replace), not roots
+
+        # Return original expr if no fusable sub-groups were found
+        return expr, True
+
+    while True:
+        original_name = expr._name
+        expr, done = _fusion_pass(expr)
+        if done or expr._name == original_name:
+            break
+
+    return expr
+
+
+@runtime_checkable
+class Fusable(Protocol):
+    """Fusable Expr Protocol
+    An `Expr` with these methods will be treated as "fusable"
+    """
+
+    def _subgraph_dependencies(self):
+        """List of `Expr` operands or `MappedArg`
+        dependencies. Since `_block_subgraph`
+        cannot include mapped arguments explicitly,
+        IO expressions must represent these task
+        arguments as `MappedArg` dependencies.
+        """
+        raise NotImplementedError
+
+    def _block_subgraph(self):
+        """Return the subgraph for an abstract 'block'.
+        Used to initialize a `SubgraphCallabe` object.
+        """
+        raise NotImplementedError
+
+
+class FusedExpr(Expr):
+    @property
+    def exprs(self):
+        return self.operands[0]
```

Comment:
You're doing this because it's hard to define `_parameters` here I'm guessing?

Reply:
Sorry, this was a debugging leftover. This can be replaced with `_parameters = ["exprs"]`